time-series-analyzer-template/src/agent.py

```diff
@@ -0,0 +1,208 @@
+        df_timeseries = df_timeseries[df_timeseries["createdAt"] < df_timeseries["createdAt"].max()]  # this row is what we want to assess against the model, so we discard
+
+        df_timeseries.rename(columns={'createdAt': 'ds', 'hash': 'y'}, inplace=True)
+        df_timeseries['ds'] = df_timeseries['ds'].dt.tz_localize(None)
+
+        # fill in missing values with median
+        median = df_timeseries['y'].median()
+        logging.info(f"Median is {median}.")
+        current_date = start_date - timedelta(minutes=start_date.minute % BUCKET_WINDOW_IN_MINUTES,
+                                              seconds=start_date.second,
+                                              microseconds=start_date.microsecond)
+        current_date += timedelta(minutes=5)
+
+        # first ensure we have values that span start to end date
+        count = 0
+        while(current_date < end_date - timedelta(minutes=BUCKET_WINDOW_IN_MINUTES)):
+            if pd.Timestamp(current_date) not in df_timeseries['ds'].values:
+                count += 1
+                df_timeseries = pd.concat([df_timeseries, pd.DataFrame({'ds': current_date, 'y': median}, index=[df_timeseries.index.max() + 1])])
+            current_date = current_date + timedelta(minutes=BUCKET_WINDOW_IN_MINUTES)
+        logging.info(f"Filled in {count} values.")
+
+        # for any values we do have that are 0, replace with median
+        logging.info(f"Replaced {len(df_timeseries[df_timeseries['y'] == 0])} values with median.")
+        df_timeseries.replace(0, df_timeseries['y'].median(), inplace=True)
+
+        m = Prophet(interval_width=INTERVAL_WIDTH)
+        m.fit(df_timeseries)
+        future = m.make_future_dataframe(periods=1, freq=str(BUCKET_WINDOW_IN_MINUTES) + 'min')
+        model = m.predict(future)
+        logging.info("Built model.")
+
+        current_value = df_current_value["hash"].iloc[0]
+        forecast = model[model["ds"] == df_current_value["createdAt"].iloc[0]]
+        yhat = forecast["yhat"].iloc[0]
+        yhat_lower = forecast["yhat_lower"].iloc[0]
+        yhat_upper = forecast["yhat_upper"].iloc[0]
+        logging.info(f"Forecast: yhat={yhat}, yhat_lower={yhat_lower}, yhat_upper={yhat_upper}; current_value={current_value}")
+
+        finding_type = get_finding_type(df_bot_alerts.iloc[0]["findingType"])
+        finding_severity = get_finding_severity(df_bot_alerts.iloc[0]["severity"])
+        if df_current_value["createdAt"].iloc[0] not in ALERTED_TIMESTAMP:
+            update_alerted_timestamp(df_current_value["createdAt"].iloc[0])
+            if current_value > yhat_upper:
+                logging.info(f"Alert detected for {CONTRACT_ADDRESS}")
+                FINDINGS_CACHE.append(TimeSeriesAnalyzerFinding.breakout("Upside", yhat, yhat_upper, current_value, CONTRACT_ADDRESS, BOT_ID, ALERT_NAME, finding_type, finding_severity))
+            if current_value < yhat_lower and current_value != 0:
```

Comment:
why not alert when `current_value=0`? Wouldn't it be useful if there's a significant drop (e.g. high yhat_lower to 0) or would it be too noisy?

Reply:
this is primarily because of the outages we see on some alerts. Ill add a comment to make this more clear and once we have more reliable bots/nodes, we can remove the condition.